doc/adr/0003-sharing-prod-debugging-cek.md

```diff
@@ -0,0 +1,95 @@
+
+- we must not compromise the performance of the production machine, and
+- the debugging machine must behave the same as the production machine.
+
+There are tradeoffs between these two requirements. If we have a separate machine, the performance of the production machine will be untouched. But there is more scope for us to make mistakes with the new machine.
+
+But if we share code between the two machines, the performance of the production machine may be compromised.
+
+This ADR proposes approaches for the two machines to share code while not compromising performance.
+
+## Decision: Polymorphic compute/return steps
+
+As long as the debugging machine has the same type, we can alter `computeCek`/`returnCek` to be polymorphic over a type-level `Bool` specifying if we’re in debug mode or not. Then we demote it to the term level in the definition of `computeCek`/`returnCek` and branch on the `Bool` thus implementing different logic depending on whether we're in debug mode or not. This promotion to the type level allows us to statically instantiate the `Bool` in an instance and thus make GHC compile the whole worker of the CEK machine twice: once in debug mode and once in production mode. Theoretically, GHC will take care to remove all the dead debug code when in production mode.
+
+### Detailed description with code snippets
+
+Whether we are debugging or not, we will be returning a `CekState`:
+
+```haskell
+data CekState uni fun =
+    -- the next state is computing
+    Computing WordArray (Context uni fun) (Closure uni fun)
+    -- the next state is returning
+    | Returning WordArray (Context uni fun) (CekValue uni fun)
+    -- evaluation finished
+    | Terminating (Term NamedDeBruijn uni fun ())
+
+data Closure uni fun = 
+  Closure (Term NamedDeBruijn uni fun ()) (CekValEnv uni fun)
+```
+
+The computing step is abstracted out to `computeCekStep` with the following signature:
+
+```haskell
+computeCekStep
+    :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => WordArray
+    -> Context uni fun
+    -> Closure uni fun
+    -> CekM uni fun s (CekState uni fun)
+```
+
+```haskell
+enterComputeCek :: forall uni fun . Bool -> (Term NamedDeBruijn uni fun ()) -> CekM uni fun s (CekState uni fun)
+enterComputeCek debug term = computeCekStep term where
+    computeCek = if debug then computeCekDebug else computeCekStep
+    {-# NOINLINE computeCek #-}  -- Making sure the `if` is only evaluated once.
+
+    computeCekDebug 
+        :: Term NamedDeBruijn uni fun ()
+        -- | The `SteppableCekM` is the monad for the debugging CEK machine. It may be just `CekM`.
+        -> SteppableCekM uni fun s (CekState uni fun) 
+    computeCekDebug term = do
+        -- `step` is the function to do a computation step. The first step is always computing the given term with an empty closure, nothing spent on the budget etc.
+        state <- step (Computing (toWordArray 0) NoFrame (Closure term Env.empty))
```

Comment:
No, `computeCekStep` serves at the stepping function! It does one step for the debugging machine and full evaluation for the production machine, it's the whole point of this representation. It's not some other `step` function.

Reply:
`computeCekStep` matches on the term and calls `computeCek` on a subterm. It's `computeCek` that is defined differently depending on whether we're in the debug mode or not. In the debug mode `computeCek` immediately returns the current `CekState` and halts execution. In the production mode `computeCek` calls the original `computeCekStep`, i.e. in the production mode `computeCekStep` calls itself through the thin `computeCek` wrapper thus achieving recursion and replicating the old behavior of the CEK machine.

Is this a clearer description of the idea?

The choice of names is arbitrary and do feel free to call things differently.

Reply:
Yes, thank you very much. I'll fix it up tomorrow.